trie-db/src/node.rs

```diff
@@ -257,20 +282,46 @@ pub enum NodeOwned<H> {
 	Extension(NibbleVec, NodeHandleOwned<H>),
 	/// Branch node; has slice of child nodes (each possibly null)
 	/// and an optional immediate node data.
-	Branch([Option<NodeHandleOwned<H>>; nibble_ops::NIBBLE_LENGTH], Option<ValueOwned<H>>),
+	Branch(ChildrenNodesOwned<H>, Option<ValueOwned<H>>),
 	/// Branch node with support for a nibble (when extension nodes are not used).
-	NibbledBranch(
-		NibbleVec,
-		[Option<NodeHandleOwned<H>>; nibble_ops::NIBBLE_LENGTH],
-		Option<ValueOwned<H>>,
-	),
+	NibbledBranch(NibbleVec, ChildrenNodesOwned<H>, Option<ValueOwned<H>>),
 	/// Node that represents a value.
 	///
 	/// This variant is only constructed when working with a [`crate::TrieCache`]. It is only
 	/// used to cache a raw value.
 	Value(Bytes, H),
 }
 
+/// Owned children of a node.
+///
+/// The maximum size of the children is `nibble_ops::NIBBLE_LENGTH`, in order to save memory we
+/// represent only the children until the last missing one. For example, if the children are
+/// `[Some(_), None, Some(_), None, None, None, None, None None, None, None, None, None, None, None,
+/// None]` we will only store in the vector `[Some(_), None, Some(_)]`.
+///
+/// This is useful to save space in the TrieCache when the children are not fully populated.
+///
+/// Before this struct was introduced, the children were stored in a fixed size array of length
+/// `nibble_ops::NIBBLE_LENGTH`, to make sure there aren't any direct accesses to the array that
+/// we've missed, wrap in its on struct, rather than directly use the `Vec`.
+#[derive(Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct ChildrenNodesOwned<H>(Vec<Option<NodeHandleOwned<H>>>);
+
+impl<H> ChildrenNodesOwned<H> {
+	pub fn get(&self, index: usize) -> Option<&Option<NodeHandleOwned<H>>> {
+		self.0.get(index)
```

Comment:
```suggestion
	pub fn get(&self, index: usize) -> Option<&NodeHandleOwned<H>> {
		self.0.get(index).map_or(None, |v| v.as_ref())
```

If a entry doesn't exist, it means it was `None`. Thus, we can simplify it.

Reply:
Some docs for the functions would also be nice.